drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+						2);
+				} else {
+					/* Mixed bus with I2C FM+ device */
+					free_timing = (uint8_t)ceil(
+						(STM32_I3C_TBUF_FMP_MIN_NS * i3c_clock / 1e9 -
+						 0.5) /
+						2);
+				}
+			} else {
+				return -EINVAL;
+			}
+		} else {
+			/* Pure I3C bus */
+			free_timing =
+				(uint8_t)ceil((STM32_I3C_TCAS_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+		}
+	}
+
+	aval = DIV_ROUND_UP(1000ull * i3c_clock, 1000000000ull) - 1;
+
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
+	LL_I3C_SetFreeTiming(i3c, free_timing);
+	LL_I3C_SetControllerActivityState(i3c, LL_I3C_OWN_ACTIVITY_STATE_0);
+	LL_I3C_SetAvalTiming(i3c, aval);
+
+	LOG_DBG("TimingReg1 = 0x%08x", LL_I3C_GetCtrlBusCharacteristic(i3c));
+
+	return 0;
+}
+
+/* Configures the I3C module in controller mode */
+static int i3c_stm32_configure(const struct device *dev, enum i3c_config_type type, void *cfg)
+{
+	int ret;
+
+	if (type == I3C_CONFIG_TARGET || type == I3C_CONFIG_CUSTOM) {
+		return -EOPNOTSUPP;
+	}
+
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+
+	I3C_TypeDef *i3c = config->i3c;
+
+	ret = i3c_stm32_activate(dev);
+	if (ret != 0) {
+		LOG_ERR("Clock and GPIO could not be initialized for the I3C module, err=%d", ret);
+		return ret;
+	}
+
+	/* I3C Initialization */
+	LL_I3C_SetMode(i3c, LL_I3C_MODE_CONTROLLER);
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
+	LL_I3C_SetControllerActivityState(i3c, LL_I3C_OWN_ACTIVITY_STATE_0);
+
+	ret = i3c_stm32_config_clk_wave(dev);
+	if (ret != 0) {
+		LOG_ERR("TimigReg0 timing could not be calculated, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_config_ctrl_bus_char(dev);
+	if (ret != 0) {
+		LOG_ERR("TimingReg1 timing could not be calculated, err=%d", ret);
+		return ret;
+	}
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableHJAck(i3c);
+#endif
+
+	/* Configure FIFO */
+	LL_I3C_SetRxFIFOThreshold(i3c, LL_I3C_RXFIFO_THRESHOLD_1_4);
+	LL_I3C_SetTxFIFOThreshold(i3c, LL_I3C_TXFIFO_THRESHOLD_1_4);
+	LL_I3C_EnableControlFIFO(i3c);
+	LL_I3C_EnableStatusFIFO(i3c);
+
+	/* Configure Controller */
+	LL_I3C_SetOwnDynamicAddress(i3c, 0);
+	LL_I3C_EnableOwnDynAddress(i3c);
+	LL_I3C_SetStallTime(i3c, 0x00);
+	LL_I3C_DisableStallACK(i3c);
+	LL_I3C_DisableStallParityCCC(i3c);
+	LL_I3C_DisableStallParityData(i3c);
+	LL_I3C_DisableStallTbit(i3c);
+	LL_I3C_DisableHighKeeperSDA(i3c);
+
+	LL_I3C_Enable(i3c);
+
+	LL_I3C_EnableIT_FC(i3c);
+	LL_I3C_EnableIT_CFNF(i3c);
+	LL_I3C_EnableIT_SFNE(i3c);
+	LL_I3C_EnableIT_RXFNE(i3c);
+	LL_I3C_EnableIT_TXFNF(i3c);
+	LL_I3C_EnableIT_ERR(i3c);
+	LL_I3C_EnableIT_WKP(i3c);
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableIT_IBI(i3c);
+	LL_I3C_EnableIT_HJ(i3c);
+#endif
+
+	/* Bus will be idle initially */
+	data->msg_state = STM32_I3C_MSG_IDLE;
+	data->sf_state = STM32_I3C_SF_IDLE;
+	data->target_id = 0;
+#ifdef CONFIG_I3C_USE_IBI
+	data->ibi_payload = 0;
+	data->ibi_payload_size = 0;
+	data->ibi_target_addr = 0;
+#endif
```

Comment:
I would expect these to be in the `init` function, not in the configure function